src/vqvae/vqvae.py

```diff
@@ -0,0 +1,188 @@
+        x = self._conv_1(inputs)
+        x = F.relu(x)
+        
+        x = self._conv_2(x)
+        x = F.relu(x)
+        
+        x = self._conv_3(x)
+        return self._residual_stack(x)
+
+
+class Decoder(nn.Module):
+    def __init__(self, in_channels, num_hiddens, num_residual_layers, num_residual_hiddens):
+        super(Decoder, self).__init__()
+        
+        self._conv_1 = nn.Conv2d(in_channels=in_channels,
+                                 out_channels=num_hiddens,
+                                 kernel_size=3, 
+                                 stride=1, padding=1)
+        
+        self._residual_stack = ResidualStack(in_channels=num_hiddens,
+                                             num_hiddens=num_hiddens,
+                                             num_residual_layers=num_residual_layers,
+                                             num_residual_hiddens=num_residual_hiddens)
+        
+        self._conv_trans_1 = nn.ConvTranspose2d(in_channels=num_hiddens, 
+                                                out_channels=num_hiddens//2,
+                                                kernel_size=4, 
+                                                stride=2, padding=1)
+        
+        self._conv_trans_2 = nn.ConvTranspose2d(in_channels=num_hiddens//2, 
+                                                out_channels=3,
+                                                kernel_size=4, 
+                                                stride=2, padding=1)
+
+    def forward(self, inputs):
+        x = self._conv_1(inputs)
+        
+        x = self._residual_stack(x)
+        
+        x = self._conv_trans_1(x)
+        x = F.relu(x)
+        
+        return self._conv_trans_2(x)
+
+
+
+class VQVAE(pl.LightningModule):
+    
+    def __init__(self):
+        super().__init__()
+        self.encoder = Encoder(
+            in_channels=3,
+            num_hiddens=128,
+            num_residual_layers=2,
+            num_residual_hiddens=32,
+        )
+        self.conv_emb = nn.Conv2d(
+            in_channels=128,
+            out_channels=128,
+            kernel_size=1,
+            stride=1)
+        self.embeddings = nn.Embedding(512, 128)
+        self.embeddings.weight.data.uniform_(-1/512, 1/512)
+        self.decoder = Decoder(
+            in_channels=128,
+            num_hiddens=128,
+            num_residual_layers=2,
+            num_residual_hiddens=32,
+        )
+
+    def forward(self, inputs):
+        z_e = self.encoder(inputs)  # (b, c, h, w)
+        z_e = self.conv_emb(z_e)
+        z_e = z_e.permute((0, 2, 3, 1))  # (b, h, w, c)
+
+        sq_dist = (z_e[:, :, :, None, :] - self.embeddings.weight).pow(2).sum(dim=-1)  # (b, h, w, 512)
+        e_index = sq_dist.argmin(dim=-1)  # (b, h, w) 
+        e = self.embeddings(e_index)  # (b, h, w, c)
+
+        z_q = z_e + (e - z_e).detach()  # Copy gradient from z_q to z_e  (b, h, w, c)
+
+        z_q = z_q.permute((0, 3, 1, 2))  # (b, c, h, w)
```

Comment:
Just so I know what is going on here, why permuting here?

Reply:
And on the 2 lines below

Reply:
Just to move the channel dimension back into its original place